### PR TITLE
Speed up word2vec model loading

### LINKED ISSUE #2642: Speed up word2vec / fasttext model loading
Loading a large word2vec model with `load_word_format(binary=True)` is slow. Users complain that loading the "standard" models published by Facebook is too slow, plus it also affects the speed of our own tests and tutorial autogeneration.

Some numbers:

```python
time gensim.models.keyedvectors.KeyedVectors.load_word2vec_format('./word2vec-google-news-300.gz', binary=True)
2019-10-21 22:24:08,326 : INFO : loading projection weights from ./word2vec-google-news-300.gz
2019-10-21 22:26:54,620 : INFO : loaded (3000000, 300) matrix from ./word2vec-google-news-300.gz
CPU times: user 2min 42s, sys: 3.64 s, total: 2min 46s
Wall time: 2min 46s
```

The I/O part itself = only loading the bytes from the file without any interpretation, takes about 30 seconds:

```python
time full = io.BytesIO(smart_open.open('./word2vec-google-news-300.gz', 'rb').read())
CPU times: user 20.9 s, sys: 8.13 s, total: 29.1 s
Wall time: 31.9 s
```

…which means our parsing code is taking up the majority of the `load_word2vec_format` time. Ideally, we shouldn't need much more than the 30 seconds (= the raw I/O speed) for the full `load_word_format(binary=True)`. Nearly 3 minutes is too much.

**Task**: Optimize `load_word2vec_format`, especially the `binary=True` branch. The code seems to live here:
https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/utils_any2vec.py#L369

### PR DESCRIPTION
**Pull request by lopusz**
Fixes #2642 .

### CODE REVIEW

File: gensim/models/utils_any2vec.py [hunk 1]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
     datatype : type, optional
         (Experimental) Can coerce dimensions to a non-default float type (such as `np.float16`) to save memory.
         Such types may result in much slower bulk operations or incompatibility with optimized routines.)
+    binary_chunk_size : int, optional
+        Size of chunk in which binary files are read. Used mostly for testing. Defalut value 100 kB.
```
Comment (piskvorky) [inline gensim/models/utils_any2vec.py]:
```suggestion
        Read input file in chunks of this many bytes, for performance reasons.
```

Does this need to be an API parameter (as opposed to a constant)? Under what conditions would a user want to change this parameter?

File: gensim/models/utils_any2vec.py [hunk 2]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+            # use count from the vocab file
+            result.vocab[word] = Vocab(index=word_id, count=counts[word])
+        else:
+            # vocab file given, but word is missing -- set count to None (TODO: or raise?)
+            logger.warning("vocabulary file is incomplete: '%s' is missing", word)
+            result.vocab[word] = Vocab(index=word_id, count=None)
+        result.vectors[word_id] = weights
+        result.index2word.append(word)
+
+    def __remove_initial_new_line(s):
```
Comment (piskvorky) [inline gensim/models/utils_any2vec.py]: Difference to `s.lstrip('\n')`?

File: gensim/models/utils_any2vec.py [hunk 3]
```diff
@@ -192,7 +242,6 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
             for line in fin:
                 word, count = utils.to_unicode(line, errors=unicode_errors).strip().split()
                 counts[word] = int(count)
-
```
Comment (piskvorky) [inline gensim/models/utils_any2vec.py]: Keep the blank line please. It helps separate code blocks for readability.

File: gensim/models/utils_any2vec.py [hunk 4]
```diff
@@ -202,43 +251,21 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
-                add_word(word, weights)
+            chunk = b''
+            tot_processed_words = 0
+
+            while tot_processed_words < vocab_size:
+                new_chunk = fin.read(binary_chunk_size)
+                chunk += new_chunk
+                max_words = vocab_size - len(result.vocab)
+                processed_words, chunk = __add_words_from_binary_chunk_to_result(result, counts, max_words,
+                                                                chunk, vector_size, datatype)
```
Comment (piskvorky) [inline gensim/models/utils_any2vec.py]: Hanging indent please (here and everywhere). No vertical indent in Gensim.

File: gensim/models/utils_any2vec.py [hunk 5]
```diff
@@ -202,43 +251,21 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+                tot_processed_words += processed_words
+                if len(new_chunk) < binary_chunk_size:
+                    break
+            if tot_processed_words != vocab_size:
+                raise EOFError("unexpected end of input; is count incorrect or file otherwise damaged?")
```
Comment (piskvorky) [inline gensim/models/utils_any2vec.py]: This seems too harsh.

If the user requests top 30k vocab, but the input file has only 20k word vectors in it, we shouldn't fail with an exception. They have no way of knowing how many words are inside a-priori.

We want to act like Python `[: n]` slicing, like in the original code: return the vocab prefix that exists.

Comment (piskvorky) [reply gensim/models/utils_any2vec.py]: Scratch that - I mixed up `limit` with `vocab_size`. The code still acts like `[: limit]`, right?

File: gensim/models/utils_any2vec.py [hunk 6]
```diff
@@ -202,43 +251,21 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
-                # use count from the vocab file
-                result.vocab[word] = Vocab(index=word_id, count=counts[word])
-            else:
-                # vocab file given, but word is missing -- set count to None (TODO: or raise?)
-                logger.warning("vocabulary file is incomplete: '%s' is missing", word)
-                result.vocab[word] = Vocab(index=word_id, count=None)
-            result.vectors[word_id] = weights
-            result.index2word.append(word)
-
         if binary:
```
Comment (mpenkov) [inline gensim/models/utils_any2vec.py]: While we're doing this, I think it's worth to split out these things into separate functions, for readability:

```
if binary:
    _word2vec_read_binary(fin, ...)
else:
    _word2vec_read_text(fin, ...)
```

What do you guys think?

File: gensim/test/test_utils_any2vec.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2017 Radim Rehurek <me@radimrehurek.com>
```
Comment (mpenkov) [inline gensim/test/test_utils_any2vec.py]:
```suggestion
# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
```

File: gensim/models/utils_any2vec.py [hunk 8]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
         Returns the loaded model as an instance of :class:`cls`.
 
     """
+
+    def __add_word_to_result(result, counts, word, weights):
+        word_id = len(result.vocab)
+        if word in result.vocab:
+            logger.warning("duplicate word '%s' in %s, ignoring all but first", word, fname)
+            return
+        if counts is None:
```
Comment (mpenkov) [inline gensim/models/utils_any2vec.py]: You're making the reader do a bit more reading then necessary. You could save your future readers some eyeball work by doing:

```python
if counts is None:
    # most common scenario: no vocab file given. just make up some bogus counts, in descending order
    word_count = vocab_size - word_id
elif word in counts:
    # use count from the vocab file
    word_count = counts[word]
else:
    logger.warning("vocabulary file is incomplete: '%s' is missing", word)
    word_count = None

result.vocab[word] = Vocab(index=word_id, count=word_count)
```

If it does not affect performance, you could also replace the elif-else with a try-catch, as that'd be more Pythonic.

File: gensim/models/utils_any2vec.py [hunk 9]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+        i = 0
+        while i < len(s) and s[i] == '\n':
+            i += 1
+        return s[i:]
+
+    def __add_words_from_binary_chunk_to_result(result, counts, max_words, chunk, vector_size, datatype):
+        start = 0
+        n = len(chunk)
+        processed_words = 0
+        n_bytes_per_vector = vector_size * dtype(REAL).itemsize
```
Comment (mpenkov) [inline gensim/models/utils_any2vec.py]: Shorter name, same meaning.

```suggestion
        bytes_per_vector = vector_size * dtype(REAL).itemsize
```

File: gensim/models/utils_any2vec.py [hunk 10]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+
+    def __remove_initial_new_line(s):
+        i = 0
+        while i < len(s) and s[i] == '\n':
+            i += 1
+        return s[i:]
+
+    def __add_words_from_binary_chunk_to_result(result, counts, max_words, chunk, vector_size, datatype):
+        start = 0
+        n = len(chunk)
```
Comment (mpenkov) [inline gensim/models/utils_any2vec.py]: I think `n` is a bit too cryptic here (it also clashes a bit with `n_bytes_per_vector` (is this intentional?).

```suggestion
        n = len(chunk)
```

You could call it `len_chunk` but I'm not sure if there's much benefit to using that over just `len(chunk)`.

File: gensim/models/utils_any2vec.py [hunk 11]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+        processed_words = 0
+        n_bytes_per_vector = vector_size * dtype(REAL).itemsize
+
+        for _ in range(0, max_words):
+            i_space = chunk.find(b' ', start)
+            i_vector = i_space + 1
+            if i_space != -1 and (n - i_vector) >= n_bytes_per_vector:
```
Comment (mpenkov) [inline gensim/models/utils_any2vec.py]: I think it'd be better for readability if you did:

```
if termination_condition:
    break

word = ...
vector = frombuffer(...)
# and so on
```

File: gensim/models/utils_any2vec.py (hunk 1 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: This parameter was useful for testing. I used it in unit tests e.g. to make sure that code is correct in exotic cases like chunk being equal to word+vector block size, etc.

I did not surface it to more "user facing" functions such as `keyedvectors.KeyedVectors.load_word2vec_format`

File: gensim/models/utils_any2vec.py (hunk 2 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Good catch! For some queer reason i forgot that except strip there is also lstrip. Corrected.

File: gensim/models/utils_any2vec.py (hunk 3 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Corrected.

File: gensim/models/utils_any2vec.py (hunk 4 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Corrected!

File: gensim/models/utils_any2vec.py (hunk 5 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Yes. The code "glues" `limit` and `vocab_size` doing  
```
vocab_size = min(vocab_size, limit)
```
(This was the original behaviour)

`vocab_size` is known from header, so file should contain at least as `vocab_size` vectors (even if user specifies strange limit).
 
There is a test  in `gensim` which check if this function throws an exception in the too short file case

https://github.com/RaRe-Technologies/gensim/blob/3d6596112f8f1fc0e839a32c5a00ef3d7365c264/gensim/test/test_word2vec.py#L426

File: gensim/models/utils_any2vec.py [hunk 12]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+            i += 1
+        return s[i:]
+
+    def __add_words_from_binary_chunk_to_result(result, counts, max_words, chunk, vector_size, datatype):
+        start = 0
+        n = len(chunk)
+        processed_words = 0
+        n_bytes_per_vector = vector_size * dtype(REAL).itemsize
+
+        for _ in range(0, max_words):
```
[the remark this answers is not in the document]
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Corrected.

File: gensim/models/utils_any2vec.py [hunk 13]
```diff
@@ -176,14 +176,64 @@ def _load_word2vec_format(cls, fname, fvocab=None, binary=False, encoding='utf8'
...
+        result.vectors[word_id] = weights
+        result.index2word.append(word)
+
+    def __remove_initial_new_line(s):
+        i = 0
+        while i < len(s) and s[i] == '\n':
+            i += 1
+        return s[i:]
+
+    def __add_words_from_binary_chunk_to_result(result, counts, max_words, chunk, vector_size, datatype):
```
[the remark this answers is not in the document]
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Simplified.

File: gensim/models/utils_any2vec.py (hunk 6 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: I think this is a good change. Refactored binary and text parts to a separate functions.

File: gensim/test/test_utils_any2vec.py (hunk 7 above)
Comment (lopusz) [reply gensim/test/test_utils_any2vec.py]: Corrected.

File: gensim/models/utils_any2vec.py (hunk 9 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Corrected.

File: gensim/models/utils_any2vec.py (hunk 10 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Good catch. Changed to `len(chunk)` call.

File: gensim/models/utils_any2vec.py (hunk 11 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: Definitely agree. Corrected.

File: gensim/models/utils_any2vec.py (hunk 8 above)
Comment (lopusz) [reply gensim/models/utils_any2vec.py]: This was unchanged by my patches, however, your version of the code definitely reads better.
Included.

**Comment by lopusz** [comment]
Hi Radim, Hi Michael,

thank you for very quick & detailed review and + a lot of  suggestions.
I cleaned up the code according to your hints.

Do you think this is now mergeable?

Best regards,
Michał

**Comment by piskvorky** [comment]
I wish all our PRs were this nice and clean! 🌟

**Pull request merged**